test/upgradeableProxy/Dispatcher.upgrade.t.sol

```diff
@@ -0,0 +1,353 @@
+        // // Deploy old version
+        string memory oldPortPrefix = "Oldpolyibc.eth.";
+        OptimisticLightClient oldConsensusManager = new OptimisticLightClient(1, oldOpProofVerifier, l1BlockProvider);
+        (dispatcher,) = TestUtils.deployDispatcherProxyAndImpl(oldPortPrefix, oldConsensusManager);
+
+        // Upgrade dispatcher for tests
+        impl = new DispatcherV2Initializable();
+        // string memory portPrefix = "polyibc.eth1.";
+        bytes memory initData = abi.encodeWithSignature("initialize(string,address)", portPrefix, consensusStateManager);
+        UUPSUpgradeable(address(dispatcher)).upgradeToAndCall(address(impl), initData);
+    }
+}
+
+contract DispatcherUpgradeTest is DispatcherUpdateClientTestSuite, DispatcherIbcWithRealProofsSuite {
+    function setUp() public virtual override {
+        consensusStateManager = new OptimisticLightClient(1, opProofVerifier, l1BlockProvider);
+        (dispatcher,) = UpgradeTestUtils.deployAndUpgradeDispatcher(
+            consensusStateManager, opProofVerifier, l1BlockProvider, "polyibc.eth1."
+        );
+        mars = new Mars(dispatcher);
+    }
+}
+
+contract DispatcherUpgradeChannelHandshakeTest is ChannelHandshakeTestSuite {
+    function setUp() public override {
+        (dispatcher,) = UpgradeTestUtils.deployAndUpgradeDispatcher(
+            dummyConsStateManager, opProofVerifier, l1BlockProvider, "polyibc.eth1."
+        );
+        mars = new Mars(dispatcher);
+        _local = LocalEnd(mars, portId, "channel-1", connectionHops, "1.0", "1.0");
+        _remote = CounterParty("eth2.7E5F4552091A69125d5DfCb7b8C2659029395Bdf", "channel-2", "1.0");
+    }
+}
+
+// This Base contract provides an open channel for sub-contract tests
+contract ChannelOpenTestBaseSetup is Base {
+    string portId = "eth1.7E5F4552091A69125d5DfCb7b8C2659029395Bdf";
+    bytes32 channelId = "channel-1";
+    address relayer = deriveAddress("relayer");
+    bool feeEnabled = false;
+
+    LocalEnd _local;
+    CounterParty _remote;
+    Mars mars;
+
+    function setUp() public virtual override {
+        (dispatcher,) = UpgradeTestUtils.deployAndUpgradeDispatcher(
+            dummyConsStateManager, opProofVerifier, l1BlockProvider, portPrefix
+        );
+        ChannelHandshakeSetting memory setting =
+            ChannelHandshakeSetting(ChannelOrder.ORDERED, feeEnabled, true, validProof);
+
+        // anyone can run Relayers
+        vm.startPrank(relayer);
+        vm.deal(relayer, 100_000 ether);
+        mars = new Mars(dispatcher);
+
+        _local = LocalEnd(mars, portId, channelId, connectionHops, "1.0", "1.0");
+        _remote = CounterParty("eth2.7E5F4552091A69125d5DfCb7b8C2659029395Bdf", "channel-2", "1.0");
+
+        openChannel(_local, _remote, setting, true);
+        connectChannel(_local, _remote, setting, false, true);
+    }
+}
+
+contract DispatcherSendPacketTestSuite is ChannelOpenTestBaseSetup {
```

Comment:
Are these tests different from the ones in Dispatcher.t.sol?

Reply:
the test suite is the same but the setup is different; the setup does an upgrade to a "new" DispatcherV2 implementation so this tests that there aren't any storage collisions (presumably the tests should fail on the new upgrade if there are)

Reply:
Though I guess the "new" Dispatcher V2 isn't much different from the old one; maybe i can change the logic in the methods a bit, or add more variables or something? Not sure if you had any ideas on specific logic we should change to test?

Reply:
I guess if there were problems with the storage after the upgrade, you'd see them when calling a single of the dispatcher methods (ie `recvPacket`) right? Maybe the test can do the upgrade and call one or two methods. Easier than duplicating all the other tests. Would it make sense to test multiple upgrades too?

Reply:
Storage collisions can happen at any stored variable/mappings; e.g. there might be a collision in the `nextSequenceSend` mapping; so if that was the case then just testing `recvPacket` wouldn't catch it since it looks like that method only touches `nextSequence Recv`. 

My view of testing _all_ of the functionality is that it is kind of like regression testing; there is an upgrade to the code, so we should make sure all of the previous functionality is working - I guess maybe [differential testing](https://book.getfoundry.sh/forge/differential-ffi-testing) is probably the best way to do this eventually; but this might be a good substitute in the mean time. 

I also wanted to re-use this test suite to do fork testing before we deploy upgrades; just to be extra sure that the contracts in use by the prod testnet won't be negatively impacted by upgrades

The test suite runs fairly quickly (< 100ms) so I don't think there's a downside to running all of the tests on the upgrade; was your concern mainly over bloating the test suite with too much code? 

Now that I'm thinking about this again, I'm realizing maybe we should also store non-trivial state in the contract before we upgrade; to really test that that state is not impacted